clientcredentialsx/auth.go

```diff
@@ -0,0 +1,70 @@
+package clientcredentialsx
+
+import (
+	"context"
+	"net/http"
+
+	"github.com/spf13/pflag"
+	"github.com/spf13/viper"
+
+	"go.infratographer.com/x/viperx"
+
+	"golang.org/x/oauth2/clientcredentials"
+)
+
+// AuthConfig is an oauth2 client config
+type AuthConfig struct {
+	// ClientID is the applications's ID
+	ClientID string
+
+	// ClientSecret is the application's secret
+	ClientSecret string
+
+	// TokenURL is the resource server's token endpoint URL. This is a constant specific to each server.
+	TokenURL string
+
+	// Scopes specifies optional requested permissions
+	Scopes []string
+}
+
+// Auth handles client credentials oauth2 authentication
+type Auth struct {
+	cfg clientcredentials.Config
+}
+
+// NewAuth returns a new oauth2 http client
+func NewAuth(cfg AuthConfig) *Auth {
+	ccCfg := clientcredentials.Config{
+		ClientID:     cfg.ClientID,
+		ClientSecret: cfg.ClientSecret,
+		TokenURL:     cfg.TokenURL,
+		Scopes:       cfg.Scopes,
+	}
+
+	cli := &Auth{
+		cfg: ccCfg,
+	}
+
+	return cli
+}
+
+// HTTPClient returns an http client using the configured client credentials config.
+// The token will auto-refresh as necessary.
+func (a Auth) HTTPClient(ctx context.Context) *http.Client {
+	return a.cfg.Client(ctx)
+}
+
+// MustViperFlags adds oidc oauth2 client config to the provided flagset and binds to viper
+func MustViperFlags(v *viper.Viper, flags *pflag.FlagSet) {
+	flags.Bool("oidc-client", true, "use oidc client auth")
+	viperx.MustBindFlag(v, "oidc.client.enabled", flags.Lookup("oidc-client"))
```

Comment:
There's not really a scenario where the client would be enabled or not, AFAICT.

Reply:
I think there's room for enabled/disabled. There may be some environments where someone deploying these apps may have the OIDC bits enabled/disabled and we should give folks the flexibility to flip these switches based off of what they want to run.

As an example, during my testing it's been useful to be able to enable/disable these integrations to help scope where bugs were popping up.

Reply:
in haproxy-manager, I'm keying off of an empty `oidc.client.id`